src/runtime_src/tools/xclbinutil/Section.cxx

```diff
@@ -73,85 +140,108 @@ Section::setName(const std::string &_sSectionName)
     auto errMsg = boost::format("ERROR: Attempting to register (%d : %s). Constructor enum of kind (%d) already registered.")
-                                % (unsigned int)_eKind % _sKindStr % (unsigned int)_eKind;
+                                % eKind % cmdName % eKind;
     throw std::runtime_error(errMsg.str());
   }
 
-  if (m_mapNameToId.find(_sKindStr) != m_mapNameToId.end()) {
-    auto errMsg = boost::format("ERROR: Attempting to register: (%d : %s). Constructor name '%s' already registered to eKind (%d).")
-                                % (unsigned int)_eKind % _sKindStr
-                                % _sKindStr % (unsigned int)m_mapNameToId[_sKindStr];
-    throw std::runtime_error(errMsg.str());
+  // Is the cmd name already registered
+  {
+    auto iter = std::find_if(sections.begin(), sections.end(), [&](const auto &entry) { return entry->getSectionName() == cmdName; });
+    if (iter != sections.end()) {
+      auto errMsg = boost::format("ERROR: Attempting to register: (%d : %s). Constructor name '%s' already registered to eKind (%d).")
+                                  % eKind % cmdName
+                                  % cmdName % iter->get()->getType();
+      throw std::runtime_error(errMsg.str());
+    }
   }
 
-  if (!_sHeaderJSONName.empty()) {
-    if (m_mapJSONNameToKind.find(_sHeaderJSONName) != m_mapJSONNameToKind.end()) {
+  // Is the header name already registered
+  if (!nodeName.empty()) {
+    auto iter = std::find_if(sections.begin(), sections.end(), [&](const auto &entry) { return entry->getNodeName() == nodeName; });
+    if (iter != sections.end()) {
       auto errMsg = boost::format("ERROR: Attempting to register: (%d : %s). JSON mapping name '%s' already registered to eKind (%d).")
-                                  % (unsigned int) _eKind % _sKindStr.c_str()
-                                  % _sHeaderJSONName % (unsigned int) m_mapJSONNameToKind[_sHeaderJSONName];
-      throw std::runtime_error(errMsg.str());
+                                    % eKind % cmdName
+                                    % nodeName % iter->get()->getType();
+        throw std::runtime_error(errMsg.str());
     }
-    m_mapJSONNameToKind[_sHeaderJSONName] = _eKind;
   }
 
-  
-  // At this point we know we are good, lets initialize the arrays
-  // TODO: These mappings are no longer scalable. 
-  //       Please refactor to a cleaner solution at the next earliest possibility.
-  m_mapKindToJSONName[_eKind] = _sHeaderJSONName;
-  m_mapIdToName[_eKind] = _sKindStr;
-  m_mapNameToId[_sKindStr] = _eKind;
-  m_mapIdToCtor[_eKind] = _Section_factory;
-  m_mapIdToSubSectionSupport[_eKind] = _bSupportsSubSections;
-  m_mapIdToSectionIndexSupport[_eKind] = _bSupportsIndexing;
+  auto sectionInfo = std::make_unique<SectionInfo>(eKind, cmdName, sectionCtor);
+  sectionInfo->setNodeName(nodeName);
+  sectionInfo->setSupportsSubSections(bSupportsSubSections);
+  sectionInfo->setSupportsIndexing(bSupportsIndexing);
```

Comment:
Should SectionInfo have a  more appropriate constructor to avoid the setters?
If too many permutations, consider a struct argument that is filled up by user and passed to ctor?

Reply:
I have refactored the code that exposed the SectionInfo variables.